src/core/exec-invoke.c

```diff
@@ -2877,6 +2976,18 @@ static int setup_exec_directory(
                         if (r < 0)
                                 goto fail;
                 }
+
+                /* Apply storage quotas if needed */
+                QuotaLimit ql;
+                ql = context->directories[type].exec_quota;
+
+                if (IN_SET(type, EXEC_DIRECTORY_STATE, EXEC_DIRECTORY_CACHE, EXEC_DIRECTORY_LOGS) && ql.quota_set) {
```

Comment:
i am pretty sure quota_set should not be checked here. i.e. if people first run a service with quota on, then reconfigure it and start it with quota now off, we should honour that.

i.e. in that case we don't need to release the projid i think but we do need to reset the limits. hence please call apply_exec_quotas() no matter what, regardless if quota_set is true or not. inside of apply_exec_quota() maintain only one shortcut: if quota_set is off and no project id is so far assigned, then don't assign one, and just exit early. hpwever, if quota is configured, or if a project id is already assigned, then go the full mile.

Reply:
and maybe move the conditionalization on the type into apply_exec_quota(), i.e. do early-exit for the runtime dir case